Userland/DevTools/UserspaceEmulator/SoftFPU.h

```diff
@@ -0,0 +1,665 @@
+            break;
+        case 6:
+            m_fpu_status_6 = (u8)tag;
+            break;
+        case 7:
+            m_fpu_status_7 = (u8)tag;
+            break;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    void fpu_set_tag(u8 index, FPU_Tag tag)
+    {
+        VERIFY(index < 8);
+        fpu_set_tag_absolute((m_fpu_stack_top + index) % 8, tag);
+    }
+
+    void set_tag_from_value_absolute(u8 index, long double val)
+    {
+        // (FP_NAN, FP_INFINITE, FP_ZERO, FP_SUBNORMAL, FP_ZERO,
+        switch (fpclassify(val)) {
+        case FP_ZERO:
+            fpu_set_tag_absolute(index, FPU_Tag::Zero);
+            break;
+        case FP_NAN:
+        case FP_INFINITE:
+        case FP_SUBNORMAL:
+            fpu_set_tag_absolute(index, FPU_Tag::Special);
+            break;
+        case FP_NORMAL:
+            fpu_set_tag_absolute(index, FPU_Tag::Valid);
+            break;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    void set_tag_from_value(u8 index, long double val)
+    {
+        set_tag_from_value_absolute((m_fpu_stack_top + index) % 8, val);
+    }
+
+    bool fpu_isnan(u8 index) const
+    {
+        return (fpu_get_tag(index) == FPU_Tag::Special) && isnan(fpu_get(index));
+    }
+
+    bool fpu_is_set(u8 index) const
+    {
+        return fpu_get_tag_absolute((m_fpu_stack_top + index) % 8) != FPU_Tag::Empty;
+    }
+
+    RoundingMode fpu_get_round_mode() const
+    {
+        return RoundingMode(m_fpu_round_mode);
+    }
+
+    template<Arithmetic T>
+    T fpu_round(long double) const;
+    template<Arithmetic T>
+    T fpu_round_checked(long double);
+
+    template<FloatingPoint T>
+    T fpu_convert(long double) const;
+    template<FloatingPoint T>
+    T fpu_convert_checked(long double);
+
+    void fpu_set_unordered()
+    {
+        set_c0(1);
+        set_c2(1);
+        set_c3(1);
+    }
+
+    void mmx_common()
+    {
+        // every MMX instrution (excluding EMMS) sets the fpu to `all valid`
+        m_fpu_tw = 0;
+    }
+
+    // FIXME: better MMX shadow checks
+    bool m_reg_is_mmx[8] { false };
+
+    // FIXME: Maybe change the structure of this
+    // do not pack, we need pointers from this
+    union {
+        long double m_st[8] { 0 };
+        struct
+        {
+            // this should represent *packed* data not an int
+            // this should overlap with the mantissa of the long double
+            u64 value;
+            // long doubles are weird with their supposed sizes, we assume, that
+            // gcc chooses either 12 (x32) or 16 (x64) byte as its size,
+            // if neither is the case we assume for now, that they really are
+            // 10 bytes big, as they are supposed to be
+            Conditional<sizeof(long double) == 16,
+                u64,
+                Conditional<sizeof(long double) == 12,
+                    u32,
+                    u16>>
+                __high;
```

Comment:
Uh... is this clang-format's doing? :sweat_smile: This looks a bit weird, can we unroll this with a `using`?

Reply:
I formatted it this way...
and the reason for it looking weird is because it is weird.... (thank gcc for that)
> long doubles are weird with their supposed sizes

Reply:
>  can we unroll this with a using?

idk, how would you do it?